api/src/ipc.c

```diff
@@ -108,6 +108,13 @@ static void ipc_free_allocated_completed_tokens(uint32_t tokens)
 
 static int ipc_waitfor(int (*cond)(uint32_t have, uint32_t expect), uint32_t wait_tokens, uint32_t * done_tokens, uint32_t timeout_ms)
 {
+    uvisor_spin_lock(ipc_tokens_lock());
+    /* Check we are not waiting for some unallocated tokens */
+    if ((wait_tokens & *ipc_allocated_tokens()) != wait_tokens) {
+        uvisor_spin_unlock(ipc_tokens_lock());
+        return UVISOR_ERROR_INVALID_PARAMETERS;
+    }
+    uvisor_spin_unlock(ipc_tokens_lock());
```

Comment:
Holding the lock while reading both allocated and completed tokens is sufficient.